process/src/services/service.ports.ts

```diff
@@ -1,295 +1,379 @@
+      listeners.onData(chunk);
+    });
+    this._updateConnectedPortsState();
+  }
+
+  private _onError(port: string, error: Error) {
+    const listeners: Map<string, IListeners> | undefined =
+      this._listeners.get(port);
+    if (listeners === undefined) {
+      return;
     }
+    listeners.forEach((listeners: IListeners) => {
+      listeners.onError(error);
+    });
+    this._unrefPortAllListeners(port);
+  }
 
-    private _onDisconnect(port: string) {
-        const listeners: Map<string, IListeners> | undefined = this._listeners.get(port);
-        if (listeners === undefined) {
-            return;
-        }
-        listeners.forEach((listeners: IListeners) => {
-            listeners.onDisconnect();
-        });
-        this._unrefPortAllListeners(port);
+  private _onDisconnect(port: string) {
+    const listeners: Map<string, IListeners> | undefined =
+      this._listeners.get(port);
+    if (listeners === undefined) {
+      return;
     }
+    listeners.forEach((listeners: IListeners) => {
+      listeners.onDisconnect();
+    });
+    this._unrefPortAllListeners(port);
+  }
 
-    private _getConnectionsCount(port: string) {
-        const listeners: Map<string, IListeners> | undefined = this._listeners.get(port);
-        if (listeners === undefined) {
-            return 0;
-        }
-        return listeners.size;
+  private _getConnectionsCount(port: string) {
+    const listeners: Map<string, IListeners> | undefined =
+      this._listeners.get(port);
+    if (listeners === undefined) {
+      return 0;
     }
+    return listeners.size;
+  }
 
-    private _updateConnectedPortsState() {
-        clearTimeout(this._connectedPortsState.timer);
-        this._connectedPortsState.state = {};
+  private _updateConnectedPortsState() {
+    clearTimeout(this._connectedPortsState.timer);
+    this._connectedPortsState.state = {};
 
-        this._controllers.forEach((controller: ControllerSerialPort) => {
-            this._connectedPortsState.state[controller.getPath()] = {
-                connections: this._getConnectionsCount(controller.getPath()),
-                ioState: controller.getIOState()
-            };
-            controller.clearIOState();
-        });
-        if (Object.keys(this._connectedPortsState.state).length === 0) {
-            return;
-        }
-        if (this._connectedPortsState.attempts < 10) {
-            this._connectedPortsState.attempts += 1;
-            this._connectedPortsState.timer = setTimeout(() => {
-                this._sendConnectedPortsState();
-            }, 250);
-        } else {
-            this._sendConnectedPortsState();
-        }
+    this._controllers.forEach((controller: ControllerSerialPort) => {
+      this._connectedPortsState.state[controller.getPath()] = {
+        connections: this._getConnectionsCount(controller.getPath()),
+        ioState: controller.getIOState(),
+      };
+      controller.clearIOState();
+    });
```

Comment:
Could it `this._connectedPortsState.state[controller.getPath()]` be `undefined` ? Maybe some check before?

Reply:
It should not be possible, because `controller` is an instance which was created in `public refPort`, we only add it then with the port's path as key. I'll add a check anyways, in case `undefined` somehow gets through